File: Components/DevExpressVCL/Demos/ExpressSpreadSheet/Delphi/Common/ReportPreviewUnit.pas
unit ReportPreviewUnit;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms, Dialogs, BaseForm, cxGraphics, cxControls, 
  cxLookAndFeels, cxLookAndFeelPainters, dxCore, dxCoreClasses, dxHashUtils, dxSpreadSheetCore, dxSpreadSheetCoreHistory, 
  dxSpreadSheetConditionalFormatting, dxSpreadSheetConditionalFormattingRules, dxSpreadSheetClasses, dxSpreadSheetContainers, 
  dxSpreadSheetFormulas, dxSpreadSheetHyperlinks, dxSpreadSheetFunctions, dxSpreadSheetGraphics, dxSpreadSheetPrinting, 
  dxSpreadSheetTypes, dxSpreadSheetUtils, dxSpreadSheet, cxClasses, Menus, ComCtrls;

type
  TfrmPreview = class(TForm)
    ssResult: TdxSpreadSheet;
    MainMenu1: TMainMenu;
    File1: TMenuItem;
    miSaveAs: TMenuItem;
    N1: TMenuItem;
    Close1: TMenuItem;
    Options1: TMenuItem;
    miShowFormulas: TMenuItem;
    sveDialog: TSaveDialog;
    procedure miCloseClick(Sender: TObject);
    procedure FormClose(Sender: TObject; var Action: TCloseAction);
    procedure miSaveAsClick(Sender: TObject);
    procedure miShowFormulasClick(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  Preview: TfrmPreview;

implementation

{$R *.dfm}

procedure TfrmPreview.FormClose(Sender: TObject; var Action: TCloseAction);
begin
  inherited;
  Action := caFree;
  Preview := nil;
end;

procedure TfrmPreview.miCloseClick(Sender: TObject);
begin
  Close;
end;

procedure TfrmPreview.miShowFormulasClick(Sender: TObject);
var
  I: Integer;
begin
  ssResult.BeginUpdate;
  try
    for I := 0 to ssResult.SheetCount - 1 do
      if ssResult.Sheets[I] is TdxSpreadSheetTableView then
        TdxSpreadSheetTableView(ssResult.Sheets[I]).Options.ShowFormulas := bDefault;
    ssResult.OptionsView.ShowFormulas := not ssResult.OptionsView.ShowFormulas;
  finally
    ssResult.EndUpdate;
  end;
end;

procedure TfrmPreview.miSaveAsClick(Sender: TObject);
var
  AFileName: string;
begin
  sveDialog.Filter := dxSpreadSheetFormatsRepository.GetSaveDialogFilter;
  if sveDialog.Execute then
  begin
    AFileName := sveDialog.FileName;
    if ExtractFileExt(AFileName) = '' then
    begin
      if (sveDialog.FilterIndex > 0) and (sveDialog.FilterIndex < dxSpreadSheetFormatsRepository.Count) then
        AFileName := AFileName + dxSpreadSheetFormatsRepository.Items[sveDialog.FilterIndex].GetExt
      else
       AFileName := AFileName + '.xlsx';
    end;
    ssResult.SaveToFile(AFileName);
  end;
end;

end.

File: Components/DevExpressVCL/Demos/ExpressSpreadSheet/CBuilder/Common/ReportPreviewUnit.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "ReportPreviewUnit.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxControls"
#pragma link "cxGraphics"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "dxCore"
#pragma link "dxCoreClasses"
#pragma link "dxHashUtils"
#pragma link "dxSpreadSheet"
#pragma link "dxSpreadSheetClasses"
#pragma link "dxSpreadSheetConditionalFormatting"
#pragma link "dxSpreadSheetConditionalFormattingRules"
#pragma link "dxSpreadSheetContainers"
#pragma link "dxSpreadSheetCore"
#pragma link "dxSpreadSheetCoreHistory"
#pragma link "dxSpreadSheetFormulas"
#pragma link "dxSpreadSheetFunctions"
#pragma link "dxSpreadSheetGraphics"
#pragma link "dxSpreadSheetHyperlinks"
#pragma link "dxSpreadSheetPrinting"
#pragma link "dxSpreadSheetTypes"
#pragma link "dxSpreadSheetUtils"
#pragma resource "*.dfm"

TfrmPreview *Preview;
//---------------------------------------------------------------------------
__fastcall TfrmPreview::TfrmPreview(TComponent* Owner)
	: TForm(Owner)
{
}

inline __fastcall TfrmPreview::~TfrmPreview()
{
  Preview = NULL;
}

//---------------------------------------------------------------------------
void __fastcall TfrmPreview::FormClose(TObject *Sender, TCloseAction &Action)
{
  Action = caFree;
  Preview = NULL;
}
//---------------------------------------------------------------------------


void __fastcall TfrmPreview::miCloseClick(TObject *Sender)
{
  Close();
}
//---------------------------------------------------------------------------
void __fastcall TfrmPreview::miShowFormulasClick(TObject *Sender)
{
  ssResult->BeginUpdate();
  try {
	for (int i = 0; i < ssResult->SheetCount; i++) {
	  ((TdxSpreadSheetTableView*)ssResult->Sheets[i])->Options->ShowFormulas = bDefault;
	}
	ssResult->OptionsView->ShowFormulas = !ssResult->OptionsView->ShowFormulas;
  }
  __finally {
	ssResult->EndUpdate();
  }
}
//---------------------------------------------------------------------------

void __fastcall TfrmPreview::miSaveAsClick(TObject *Sender)
{
  String AFileName;
  TdxSpreadSheetCustomFormatClass AFormatClass;
  sveDialog->Filter = dxSpreadSheetFormatsRepository()->GetSaveDialogFilter();
  if (sveDialog->Execute()) {
	AFileName = sveDialog->FileName;
	if (ExtractFileExt(AFileName) == "") {
	  if (sveDialog->FilterIndex > 0)
		AFileName = AFileName + dxSpreadSheetFormatsRepository()->Exts[sveDialog->FilterIndex - 1];
	  else
		AFileName = AFileName + ".xlsx";
	}
	ssResult->SaveToFile(AFileName);
  }
}
//---------------------------------------------------------------------------

